Upload portfolio item ui

natehood96 opened the pull request:
This is a great PR

Commit:
most UI is done for component

Commit:
upload portfolio item UI completed

Commit:
backend should be working

Commit:
everything works, just adding tests

Commit:
submit portfolio works

Commit:
added tests

Commit:
finished backend and frontend

jasonccox commented on client/src/Backend.ts:
```diff
@@ -92,3 +93,21 @@ export async function registerBusiness(business: Business): Promise<void> {
 }
+
+export async function addPortfolio(portfolioItem: PortfolioItem): Promise<void> {
+  const url = `${baseUrl}/business/portfolio`
```
The route I made for getting a business is `/business/:handle`. How would you feel about making this one `/business/:handle/portfolio`?

jasonccox commented on client/src/components/FileUpload.tsx:
```diff
@@ -14,7 +14,7 @@ const FileUpload: React.FC<Props> = (props: Props) => {
   return (
     <LabeledInput
       inputType="input"
-      label="Logo"
```
Whoops, that shouldn't have been hard-coded. Nice catch.

jasonccox commented on client/src/Backend.ts:
```diff
@@ -92,3 +93,21 @@ export async function registerBusiness(business: Business): Promise<void> {
+    })
+  }
+  catch(err){
+    throw new Error('Sorry, an unexpected error occurred. Please try again later.')
```
Let's log this error before we throw a new one.

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+    } else {
+      fileBase64 = Promise.resolve(null)
+    }
+    let filePreview = await fileBase64
+    setPreviewFile(filePreview || null)
```
I think you can just do this:
```suggestion
    setPreviewFile(await fileBase64)
```

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  async function setFile(file: File | null) {
+    let fileBase64: Promise<string|null>;
```
Did you run the linter on all these changes? I think it should make this semicolon disappear.

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+        <FileUpload
+            label="Portfolio image"
+            accept={['.jpg', '.png']}
+            onChange={(file) => setFile(file)}/>
```
```suggestion
            onChange={setFile} />
```

jasonccox commented on client/src/styles/upload-portfolio-item.scss:
```diff
@@ -0,0 +1,39 @@
+section.portfolio-upload{
+  margin: auto;
+  max-width: 75rem;
+  min-width: 30rem;
+  div.text-image-block{
+    display: flex;
+    flex-wrap: wrap;
+    justify-content: space-around;
+    textarea.description{
+      flex-basis: 50%; 
+      flex-grow: 0;
+      border: none;
+      overflow: auto;
+      outline: none;
+      -webkit-box-shadow: none;
+      -moz-box-shadow: none;
+      box-shadow: none;
+      resize: none; /*remove the resize handle on the bottom right*/
+      font-family: inherit; 
+      font-size: inherit; 
+      min-height: 8rem;
+      background-color: #DEDEDE
```
All this talk of whitespace around `else if (...)`, and then this? :laughing: Haha, it's not a big deal though

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+
+const UploadPortfolioItem: React.FC <Props> = ({handle}: Props) => {
+  const [description, setDescription] = useState<string>('')
+  const [previewFile, setPreviewFile] = useState<string|null>('')
```
From what I can tell, it seems like you're usually using `null` when there is no preview file. Why not just make the initial state `null` as well?

natehood96 committed:
Update client/src/components/UploadPortfolioItem.tsx

Co-authored-by: Jason Cox <dev@jasoncarloscox.com>

natehood96 committed:
Update client/src/components/UploadPortfolioItem.tsx

Co-authored-by: Jason Cox <dev@jasoncarloscox.com>

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+    <section className='portfolio-upload'>
+      <h2>Upload Portfolio Item</h2>
+      <div className='text-image-block'>
+        <textarea 
```
How would you feel about using a `LabelledInput` here instead? That we we get a label and consistent styling.

jasonccox commented on client/src/components/UploadPortfolioItem.tsx:
```diff
@@ -0,0 +1,74 @@
+        Publish
+      </button>  
+    </section>
+  )
```
I think we should wrap all of the inputs in a `<form>` tag. Then you can use `required` on both the description and the file to make sure we can't submit without them. For example,

```tsx
<form onSubmit={handleSubmit} >
  <LabelledInput ... />
  <button type="submit" ... >Publish</button>
</form>
```

Commit:
resolved PR comments

Commit:
resolved merge conflicts

Commit:
resolved conflicts

jasonccox committed:
Fix upload item layout

jasonccox committed:
Add finishing touches to upload portfolio item UI

- Refactor Backend.portfolioItem to take advantage of request function
- Fix layout so that things line up with small width
- Make description textarea consistent with those on other pages
- Ensure a file is always sent

jasonccox committed:
Refactor POST /business/:handle/portfolio route

- Make the portfolio service responsible for getting the business id so
  that db/portfolio.ts can just worry about the portfolio table
- Ensure tables are created in the correct order
- Rename portfolio table columns to better describe their contents
- Enforce length requirements on description and file
- Add more tests
- Rename some things to be more descriptive/consistent
- Fix a few linting errors

jasonccox committed:
Make "Add Portfolio Item" button functional
